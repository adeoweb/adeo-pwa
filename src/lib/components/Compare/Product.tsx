import React, { FunctionComponent } from 'react';

import Image from 'src/lib/components/Image';
import PriceBox from 'src/lib/components/PriceBox';
import AddToCart from 'src/lib/components/ProductItem/AddToCart';
import { CustomerModalTypes } from 'src/lib/constants/customer';
import { Link } from 'src/lib/drivers';
import { TProduct } from 'src/lib/types/graphql/Product';
import getItemUrl from 'src/lib/util/getItemUrl';
import { useAppContext } from 'src/peregrine/lib/context/adeoweb/app';
import { useProductCompareContext } from 'src/peregrine/lib/context/adeoweb/productCompare';

import AddToWishlist from '../AddToWishlist';

interface IProductProps {
    product: TProduct;
    columnStyle: {
        width: string;
    };
}

const Product: FunctionComponent<IProductProps> = ({
    product,
    columnStyle
}) => {
    const [, { removeProduct }] = useProductCompareContext();
    const [, { setCustomerModal }] = useAppContext();

    const { name, small_image: smallImage, price_range: priceRange } = product;
    const itemUrl = getItemUrl(product);
    const itemName = name || '';
    const imageLabel = (smallImage && smallImage.label) || '';
    const imageUrl = (smallImage && smallImage.url) || '';

    const handleRemoveProduct = () => {
        removeProduct(product);
    };
    const handleNotLoggedIn = () => {
        setCustomerModal(CustomerModalTypes.LOG_IN);
    };

    return (
        <td data-th="Product" className="cell product info" style={columnStyle}>
            <button
                className="btn-remove btn btn-link"
                title="Remove Product"
                onClick={handleRemoveProduct}
            >
                <i className="icon-cancel" />
            </button>
            <figure className="product-item-photo">
                <Link to={itemUrl}>
                    <Image
                        resource={'small_image'}
                        alt={imageLabel}
                        src={imageUrl}
                    />
                </Link>
            </figure>
            <h2 className="product-item-name">
                <Link to={itemUrl} title={itemName}>
                    {itemName}
                </Link>
            </h2>
            <PriceBox priceRange={priceRange} />
            <div className="product-item-actions">
                <AddToCart product={product} />
                <AddToWishlist
                    product={product}
                    handleNotLoggedIn={handleNotLoggedIn}
                />
            </div>
        </td>
    );
};

export default Product;
