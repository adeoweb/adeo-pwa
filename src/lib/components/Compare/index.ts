export { default } from './Compare';
export { default as AddToCompare } from './AddToCompare';
