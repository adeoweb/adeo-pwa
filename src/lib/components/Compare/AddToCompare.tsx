import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';
import { TProduct } from 'src/lib/types/graphql/Product';
import { useProductCompare } from 'src/peregrine/lib/talons/adeoweb/Product/useProductCompare';

type TAddToCompareProps = {
    product: TProduct;
};

const AddToCompare: FunctionComponent<TAddToCompareProps> = ({ product }) => {
    const { t } = useTranslation();
    const { setProductHandler, isProductBeingCompared } = useProductCompare();

    const addProductToCompare = () => {
        setProductHandler(product);
    };

    return (
        <button
            className={`paction add-compare${
                isProductBeingCompared(product) ? ` is-being-compared` : ``
            }`}
            title={t('Add to Compare')}
            onClick={addProductToCompare}
        >
            <span>{t('Add to Compare')}</span>
        </button>
    );
};

export default AddToCompare;
