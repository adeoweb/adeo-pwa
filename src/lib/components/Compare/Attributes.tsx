import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';

import RichText from 'src/lib/components/RichText';
import { TProductInterface } from 'src/lib/types/graphql/Product';
import { TProductData } from 'src/peregrine/lib/store/reducers/adeoweb/productCompare';

interface IAttributesProps {
    productData: TProductData;
}

type TAttributeKey = keyof TProductInterface;

type TAttributeValue = {
    title: string;
    isRichText: boolean;
};

type TAttributeOptions = Partial<{ [key in TAttributeKey]: TAttributeValue }>;

type TGetAttributeValueProps = {
    productId: string;
    attributeKey: TAttributeKey;
};

type TGetAttributeValue = (
    props: TGetAttributeValueProps
) => string | React.ReactNode;

const Attributes: FunctionComponent<IAttributesProps> = ({ productData }) => {
    const { t } = useTranslation();

    if (!productData) {
        return null;
    }

    const attributes: TAttributeKey[] = [
        'sku',
        'short_description',
        'description'
    ];

    const attributeOptions: TAttributeOptions = {
        sku: {
            title: 'SKU',
            isRichText: false
        },
        short_description: {
            title: t('Description'),
            isRichText: true
        },
        description: {
            title: t('Short Description'),
            isRichText: true
        }
    };

    const getAttributeValue: TGetAttributeValue = ({
        productId,
        attributeKey
    }) => {
        const options = attributeOptions[attributeKey];
        const product = productData[productId];
        const attribute = product[attributeKey];

        if (!attribute) {
            return '';
        }

        if (
            options &&
            options.isRichText &&
            typeof attribute === 'object' &&
            'html' in attribute
        ) {
            const content = (attribute && attribute.html) || '';

            return <RichText content={content} />;
        }

        return attribute || '';
    };

    return (
        <tbody>
            {attributes.map(attributeKey => {
                const options = attributeOptions[attributeKey];
                const title = (options && options.title) || attributeKey;

                return (
                    <tr key={attributeKey}>
                        <th scope="row" className="cell label">
                            <span className="attribute label">{title}</span>
                        </th>
                        {Object.keys(productData).map(productId => (
                            <td
                                key={productId}
                                className="cell product attribute"
                            >
                                <div className="attribute value">
                                    {getAttributeValue({
                                        attributeKey,
                                        productId
                                    })}
                                </div>
                            </td>
                        ))}
                    </tr>
                );
            })}
        </tbody>
    );
};

export default Attributes;
