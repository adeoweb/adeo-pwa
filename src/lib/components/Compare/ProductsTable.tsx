import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';

import Product from 'src/lib/components/Compare/Product';
import { TProductData } from 'src/peregrine/lib/store/reducers/adeoweb/productCompare';

interface IProductsProps {
    productData: TProductData;
}

const ProductsTable: FunctionComponent<IProductsProps> = ({ productData }) => {
    const { t } = useTranslation('product');

    if (!productData) {
        return null;
    }

    const columnCount = Object.keys(productData).length + 1;
    const columnWidth = 100 / (columnCount || 1);
    const columnStyle = {
        width: `${columnWidth}%`
    };

    return (
        <tbody>
            <tr>
                <th
                    scope="row"
                    className="cell label product"
                    style={columnStyle}
                >
                    <span>{t('product')}</span>
                </th>
                {productData &&
                    Object.keys(productData).map(productId => {
                        const product = productData[productId];

                        return (
                            <Product
                                key={productId}
                                product={product}
                                columnStyle={columnStyle}
                            />
                        );
                    })}
            </tr>
        </tbody>
    );
};

export default ProductsTable;
