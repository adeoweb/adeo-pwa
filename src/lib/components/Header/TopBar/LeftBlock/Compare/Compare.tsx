import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';

import { useProductCompare } from 'src/peregrine/lib/talons/adeoweb/Product/useProductCompare';

import CompareDropdown from './CompareDropdown';

const Compare: FunctionComponent = () => {
    const { t } = useTranslation('common');
    const { productData } = useProductCompare();
    const productQuantity = Object.keys(productData).length;
    const isActive = productQuantity > 0;

    return (
        <div
            className={`dropdown compare-dropdown ${isActive ? 'active' : ''}`}
        >
            <div
                className="dropdown-toggle"
                role="button"
                data-toggle="dropdown"
                aria-haspopup="true"
                aria-expanded="false"
                data-display="static"
            >
                <i className="icon-retweet" />
                {`${t('Compare')} (${productQuantity})`}
            </div>
            <CompareDropdown />
        </div>
    );
};

export default Compare;
