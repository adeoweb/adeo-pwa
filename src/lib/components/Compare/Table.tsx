import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';

import Attributes from 'src/lib/components/Compare/Attributes';
import ProductsTable from 'src/lib/components/Compare/ProductsTable';
import { TProductData } from 'src/peregrine/lib/store/reducers/adeoweb/productCompare';

interface ITableProps {
    productData: TProductData;
}

const Table: FunctionComponent<ITableProps> = ({ productData }) => {
    const { t } = useTranslation('product');

    return (
        <div className="columns">
            <div className="column main">
                <div className="table-wrapper comparison">
                    <table className="data table table-comparison scroll">
                        <caption className="table-caption">
                            {t('Compare Products')}
                        </caption>
                        <ProductsTable productData={productData} />
                        <Attributes productData={productData} />
                    </table>
                </div>
            </div>
        </div>
    );
};

export default Table;
