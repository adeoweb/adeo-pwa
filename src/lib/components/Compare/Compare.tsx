import isEmpty from 'lodash.isempty';

import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';

import Table from 'src/lib/components/Compare/Table';
import { useCompare } from 'src/peregrine/lib/talons/adeoweb/Compare/useCompare';

const Compare: FunctionComponent = () => {
    const { t } = useTranslation('product');

    const { productData } = useCompare();

    const productDataIsValid = productData && !isEmpty(productData);

    return (
        <div className="page-main">
            <div className="page-title-wrapper">
                <h1 className="page-title">
                    <span className="base">{t('Compare Products')}</span>
                </h1>
            </div>
            {productDataIsValid ? (
                <Table productData={productData} />
            ) : (
                <div>{t('Product list is empty')}</div>
            )}
        </div>
    );
};

export default Compare;
