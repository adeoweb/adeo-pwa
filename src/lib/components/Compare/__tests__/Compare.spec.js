import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import { createTestInstance } from '@magento/peregrine';

import Compare from 'src/lib/components/Compare';
import { mockProductData } from 'src/lib/components/Compare/__tests__/mockData';

jest.mock('src/peregrine/lib/context/adeoweb/productCompare', () => {
    return {
        useProductCompareContext: () => [
            {},
            {
                removeProduct: jest.fn(() => {})
            }
        ]
    };
});

jest.mock('src/peregrine/lib/talons/adeoweb/Wishlist/useWishlistItem', () => {
    return {
        useWishlistItem: () => [
            {},
            {
                handleAddToWishlist: jest.fn(() => {}),
                handleRemoveFromWishlist: jest.fn(() => {}),
                isInWishlist: false,
                isSignedIn: false
            }
        ]
    };
});

jest.mock('src/peregrine/lib/context/adeoweb/app', () => {
    return {
        useAppContext: () => [
            {},
            {
                setCustomerModal: jest.fn(() => {})
            }
        ]
    };
});

jest.mock('src/peregrine/lib/talons/adeoweb/Compare/useCompare', () => {
    return {
        useCompare: () => ({
            productData: mockProductData
        })
    };
});

jest.mock('src/peregrine/lib/talons/adeoweb/Product/useAddToCart', () => {
    return {
        useAddToCart: () => ({
            handleAddToCart: () => {},
            isAddToCartDisabled: false
        })
    };
});
test('Footer component renders correctly', () => {
    const tree = createTestInstance(
        <Router>
            <Compare />
        </Router>
    );

    expect(tree.toJSON()).toMatchSnapshot();
});
