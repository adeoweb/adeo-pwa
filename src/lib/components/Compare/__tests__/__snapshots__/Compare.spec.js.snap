// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component renders correctly 1`] = `
<div
  className="page-main"
>
  <div
    className="page-title-wrapper"
  >
    <h1
      className="page-title"
    >
      <span
        className="base"
      >
        Compare Products
      </span>
    </h1>
  </div>
  <div
    className="columns"
  >
    <div
      className="column main"
    >
      <div
        className="table-wrapper comparison"
      >
        <table
          className="data table table-comparison scroll"
        >
          <caption
            className="table-caption"
          >
            Compare Products
          </caption>
          <tbody>
            <tr>
              <th
                className="cell label product"
                scope="row"
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                <span>
                  product
                </span>
              </th>
              <td
                className="cell product info"
                data-th="Product"
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                <button
                  className="btn-remove btn btn-link"
                  onClick={[Function]}
                  title="Remove Product"
                >
                  <i
                    className="icon-cancel"
                  />
                </button>
                <figure
                  className="product-item-photo"
                >
                  <a
                    href="/urlKey.html"
                    onClick={[Function]}
                  >
                    <div
                      className="undefined undefined"
                    >
                      <img
                        alt="small image label"
                        className="undefined undefined"
                        loading="eager"
                        src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAFCAQAAADIpIVQAAAADklEQVR42mNkgAJGIhgAALQABsHyMOcAAAAASUVORK5CYII="
                        style={
                          Object {
                            "height": "auto",
                            "width": "100%",
                          }
                        }
                      />
                      <img
                        alt="small image label"
                        className="undefined undefined"
                        loading="lazy"
                        onError={[Function]}
                        onLoad={[Function]}
                        src="smallImagetUrl"
                        style={
                          Object {
                            "height": "auto",
                            "width": "100%",
                          }
                        }
                      />
                    </div>
                  </a>
                </figure>
                <h2
                  className="product-item-name"
                >
                  <a
                    href="/urlKey.html"
                    onClick={[Function]}
                    title="product name"
                  >
                    product name
                  </a>
                </h2>
                <div
                  className="price-box"
                >
                  <span
                    className="product-price"
                  >
                    <div />
                  </span>
                </div>
                <div
                  className="product-item-actions"
                >
                  <button
                    className="btn-icon btn-add-cart"
                    disabled={false}
                    onClick={[Function]}
                  >
                    <div
                      className="add-to-cart-content"
                    >
                      <svg
                        fill="none"
                        height="24"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16.667 7.99954V6.03288C16.667 5.42004 16.5463 4.81321 16.3118 4.24702C16.0772 3.68084 15.7335 3.16639 15.3002 2.73305C14.8668 2.29971 14.3524 1.95596 13.7862 1.72144C13.22 1.48692 12.6132 1.36621 12.0003 1.36621C11.3875 1.36621 10.7807 1.48692 10.2145 1.72144C9.64828 1.95596 9.13383 2.29971 8.70049 2.73305C8.26715 3.16639 7.92341 3.68084 7.68889 4.24702C7.45437 4.81321 7.33366 5.42004 7.33366 6.03288V10.6995C7.33366 10.8764 7.4039 11.0459 7.52892 11.1709C7.65395 11.296 7.82351 11.3662 8.00033 11.3662C8.17714 11.3662 8.34671 11.296 8.47173 11.1709C8.59675 11.0459 8.66699 10.8764 8.66699 10.6995V9.33288H14.0003V7.99954H8.66699V6.03288C8.66699 5.14882 9.01818 4.30098 9.6433 3.67586C10.2684 3.05073 11.1163 2.69954 12.0003 2.69954C12.8844 2.69954 13.7322 3.05073 14.3573 3.67586C14.9825 4.30098 15.3337 5.14882 15.3337 6.03288V10.6662C15.3337 10.843 15.4039 11.0126 15.5289 11.1376C15.6539 11.2626 15.8235 11.3329 16.0003 11.3329C16.1771 11.3329 16.3467 11.2626 16.4717 11.1376C16.5968 11.0126 16.667 10.843 16.667 10.6662V9.33288H20.0003V21.3329H4.00033V9.33288H6.00033V7.99954H2.66699V21.3929C2.66699 21.7306 2.80115 22.0545 3.03994 22.2933C3.27874 22.5321 3.60262 22.6662 3.94033 22.6662H20.0603C20.398 22.6662 20.7219 22.5321 20.9607 22.2933C21.1995 22.0545 21.3337 21.7306 21.3337 21.3929V7.99954H16.667Z"
                          fill="#21293C"
                        />
                      </svg>
                      Add to cart
                    </div>
                  </button>
                  <button
                    className="btn-icon-wish"
                    onClick={[Function]}
                    title="Add to Wishlist"
                  >
                    <i
                      className="icon-heart"
                    />
                  </button>
                </div>
              </td>
            </tr>
          </tbody>
          <tbody>
            <tr>
              <th
                className="cell label"
                scope="row"
              >
                <span
                  className="attribute label"
                >
                  SKU
                </span>
              </th>
              <td
                className="cell product attribute"
              >
                <div
                  className="attribute value"
                >
                  sku-123
                </div>
              </td>
            </tr>
            <tr>
              <th
                className="cell label"
                scope="row"
              >
                <span
                  className="attribute label"
                >
                  Description
                </span>
              </th>
              <td
                className="cell product attribute"
              >
                <div
                  className="attribute value"
                >
                  
                </div>
              </td>
            </tr>
            <tr>
              <th
                className="cell label"
                scope="row"
              >
                <span
                  className="attribute label"
                >
                  Short Description
                </span>
              </th>
              <td
                className="cell product attribute"
              >
                <div
                  className="attribute value"
                >
                  <div
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "product description",
                      }
                    }
                    onClick={[Function]}
                  />
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
